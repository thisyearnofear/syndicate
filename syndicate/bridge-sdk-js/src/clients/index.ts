export * from "./evm"
export * from "./solana"
export * from "./near"
