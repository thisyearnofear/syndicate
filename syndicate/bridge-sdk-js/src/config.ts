export type NetworkType = "mainnet" | "testnet"

const ADDRESSES = {
  mainnet: {
    arb: "0xd025b38762B4A4E36F0Cde483b86CB13ea00D989",
    base: "0xd025b38762B4A4E36F0Cde483b86CB13ea00D989",
    eth: "0x3701B9859Dbb9a4333A3dd933ab18e9011ddf2C8",
    near: "omni.bridge.near",
    sol: {
      locker: "dahPEoZGXfyV58JqqH85okdHmpN8U2q8owgPUXSCPxe",
      wormhole: "worm2ZoG2kUd4vFXhvjh93UUH596ayRfgQ2MgjNMTth",
    },
  },
  testnet: {
    arb: "0x0C981337fFe39a555d3A40dbb32f21aD0eF33FFA",
    base: "0xa56b860017152cD296ad723E8409Abd6e5D86d4d",
    eth: "0x68a86e0Ea5B1d39F385c1326e4d493526dFe4401",
    near: "omni.n-bridge.testnet",
    sol: {
      locker: "862HdJV59Vp83PbcubUnvuXc4EAXP8CDDs6LTxFpunTe",
      wormhole: "3u8hJUVTA4jH1wYAyUur7FFZVQ8H635K3tSHHF4ssjQ5",
    },
  },
} as const

let selectedNetwork: NetworkType = "mainnet"

export function setNetwork(network: NetworkType) {
  selectedNetwork = network
}

export function getNetwork(): NetworkType {
  return selectedNetwork
}

export const addresses = {
  get arb() {
    return ADDRESSES[selectedNetwork].arb
  },
  get base() {
    return ADDRESSES[selectedNetwork].base
  },
  get eth() {
    return ADDRESSES[selectedNetwork].eth
  },
  get near() {
    return ADDRESSES[selectedNetwork].near
  },
  get sol() {
    return ADDRESSES[selectedNetwork].sol
  },
  get network() {
    return selectedNetwork
  },
}
