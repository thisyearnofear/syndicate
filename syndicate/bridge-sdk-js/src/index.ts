// Export main client and factory
export * from "./client"
export * from "./factory"

// Export API types and utilities
export * from "./api"
export * from "./utils"

// Export type definitions
export * from "./types"

// Export specific clients
export * from "./clients"

// Export proofs
export * from "./proofs"

export * from "./config"
