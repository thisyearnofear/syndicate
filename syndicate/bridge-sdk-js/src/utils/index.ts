export * from "./chain"
export * from "./decimals"
export * from "./tokens"
