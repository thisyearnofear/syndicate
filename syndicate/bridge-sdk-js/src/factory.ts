import type { Provider } from "@coral-xyz/anchor"
import type { WalletSelector } from "@near-wallet-selector/core"
import type { ethers } from "ethers"
import { Account } from "near-api-js"
import { EvmBridgeClient } from "./clients/evm"
import { NearBridgeClient } from "./clients/near"
import { NearWalletSelectorBridgeClient } from "./clients/near-wallet-selector"
import { SolanaBridgeClient } from "./clients/solana"
import { ChainKind } from "./types"

// Define client types
type ClientTypes = {
  [ChainKind.Near]: NearBridgeClient | NearWalletSelectorBridgeClient
  [ChainKind.Eth]: EvmBridgeClient
  [ChainKind.Base]: EvmBridgeClient
  [ChainKind.Arb]: EvmBridgeClient
  [ChainKind.Sol]: SolanaBridgeClient
}

// Define wallet types for each chain
type WalletTypes = {
  [ChainKind.Near]: Account | WalletSelector
  [ChainKind.Eth]: ethers.Signer
  [ChainKind.Base]: ethers.Signer
  [ChainKind.Arb]: ethers.Signer
  [ChainKind.Sol]: Provider
}

/**
 * Creates a chain-specific bridge client instance based on the provided chain and wallet.
 *
 * @template T - The chain kind (must be one of {@link ChainKind})
 * @param chain - The blockchain network to create a client for
 * @param wallet - Chain-specific wallet instance for signing transactions
 * @returns A strongly-typed client instance for the specified chain
 * @throws {Error} If no client implementation exists for the chain
 *
 * @example NEAR with Account
 * ```typescript
 * import { connect } from 'near-api-js';
 *
 * const nearAccount = await connect(config);
 * const client = getClient(ChainKind.Near, nearAccount);
 * // Type: NearBridgeClient
 * const txHash = await client.initDeployToken("near:token.near");
 * ```
 *
 * @example NEAR with Wallet Selector
 * ```typescript
 * import { setupWalletSelector } from '@near-wallet-selector/core';
 *
 * const selector = await setupWalletSelector({ ... });
 * const client = getClient(ChainKind.Near, selector);
 * // Type: NearWalletSelectorBridgeClient
 * const txHash = await client.initDeployToken("near:token.near");
 * ```
 *
 * @example Ethereum
 * ```typescript
 * import { ethers } from 'ethers';
 *
 * const provider = new ethers.providers.Web3Provider(window.ethereum);
 * const signer = provider.getSigner();
 * const client = getClient(ChainKind.Eth, signer);
 * // Type: EvmBridgeClient
 * const tx = await client.deposit({ to: "0x...", amount: "1.0" });
 * ```
 *
 * @example Solana
 * ```typescript
 * import { Connection, Keypair } from '@solana/web3.js';
 *
 * const connection = new Connection('https://api.mainnet-beta.solana.com');
 * const client = getClient(ChainKind.Sol, connection);
 * // Type: SolanaBridgeClient
 * const signature = await client.transfer({ ... });
 * ```
 *
 * @example Base/Arbitrum (EVM-compatible chains)
 * ```typescript
 * const baseClient = getClient(ChainKind.Base, signer);
 * const arbClient = getClient(ChainKind.Arb, signer);
 * // Both Type: EvmBridgeClient but configured for different chains
 * ```
 */

// Function overloads for each chain type
export function getClient(chain: ChainKind.Near, wallet: Account): NearBridgeClient
export function getClient(
  chain: ChainKind.Near,
  wallet: WalletSelector,
): NearWalletSelectorBridgeClient
export function getClient(chain: ChainKind.Eth, wallet: ethers.Signer): EvmBridgeClient
export function getClient(chain: ChainKind.Base, wallet: ethers.Signer): EvmBridgeClient
export function getClient(chain: ChainKind.Arb, wallet: ethers.Signer): EvmBridgeClient
export function getClient(chain: ChainKind.Sol, wallet: Provider): SolanaBridgeClient

// Generic implementation
export function getClient<T extends ChainKind>(chain: T, wallet: WalletTypes[T]): ClientTypes[T] {
  switch (chain) {
    case ChainKind.Near:
      if (wallet instanceof Account) {
        return new NearBridgeClient(wallet) as ClientTypes[T]
      }
      return new NearWalletSelectorBridgeClient(wallet as WalletSelector) as ClientTypes[T]
    case ChainKind.Eth:
    case ChainKind.Base:
    case ChainKind.Arb:
      return new EvmBridgeClient(wallet as ethers.Signer, chain) as ClientTypes[T]
    case ChainKind.Sol:
      return new SolanaBridgeClient(wallet as Provider) as ClientTypes[T]
    default:
      throw new Error(`No client implementation for chain: ${chain}`)
  }
}
