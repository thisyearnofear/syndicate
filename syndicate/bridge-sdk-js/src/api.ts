import { z } from "zod"
import { getNetwork } from "./config"
import type { OmniAddress } from "./types"

const ChainSchema = z.enum(["Eth", "Near", "Sol", "Arb", "Base"])
export type Chain = z.infer<typeof ChainSchema>

// Custom transformer for safe BigInt coercion that handles scientific notation
const safeBigInt = (nullable = false) => {
  const transformer = z.preprocess(
    (val) => {
      if (val === null && nullable) return null

      try {
        // If it's a number, convert to string without scientific notation
        if (typeof val === "number") {
          return BigInt(val.toLocaleString("fullwide", { useGrouping: false }))
        }

        // If it's a string, try direct conversion first
        if (typeof val === "string") {
          try {
            // Try direct BigInt conversion first (handles numeric strings)
            return BigInt(val)
          } catch {
            // If direct conversion fails, it might be scientific notation
            const num = Number(val)
            if (!Number.isNaN(num)) {
              return BigInt(num.toLocaleString("fullwide", { useGrouping: false }))
            }
            throw new Error(`Cannot convert ${val} to BigInt`)
          }
        }

        return val
      } catch {
        throw new Error(`Invalid BigInt value: ${val}`)
      }
    },
    nullable ? z.bigint().min(0n).nullable() : z.bigint().min(0n),
  )

  return transformer
}

// Updated based on OpenAPI spec
const NearReceiptTransactionSchema = z.object({
  block_height: z.number().int().min(0),
  block_timestamp_seconds: z.number().int().min(0),
  transaction_hash: z.string(),
})

const EVMLogTransactionSchema = z.object({
  block_height: z.number().int().min(0),
  block_timestamp_seconds: z.number().int().min(0),
  transaction_hash: z.string(),
})

// Updated to make all fields optional since we saw an empty Solana object in the example
const SolanaTransactionSchema = z.object({
  slot: z.number().int().min(0).optional(),
  block_timestamp_seconds: z.number().int().min(0).optional(),
  signature: z.string().optional(),
})

// Update to match the Transaction schema in OpenAPI spec - one of these fields will be present
const TransactionSchema = z
  .object({
    NearReceipt: NearReceiptTransactionSchema.optional(),
    EVMLog: EVMLogTransactionSchema.optional(),
    Solana: SolanaTransactionSchema.optional(),
  })
  .refine(
    (data) => {
      // Ensure exactly one of the fields is defined
      const definedFields = [data.NearReceipt, data.EVMLog, data.Solana].filter(
        (field) => field !== undefined,
      )
      return definedFields.length === 1
    },
    { message: "Exactly one transaction type must be present" },
  )

const TransferMessageSchema = z.object({
  token: z.string(),
  amount: safeBigInt(),
  sender: z.string(),
  recipient: z.string(),
  fee: z.object({
    fee: safeBigInt(),
    native_fee: safeBigInt(),
  }),
  msg: z.string().nullable(),
})

const TransfersQuerySchema = z
  .object({
    sender: z.string().optional(),
    transaction_id: z.string().optional(),
    offset: z.number().default(0),
    limit: z.number().default(10),
  })
  .refine((data) => data.sender || data.transaction_id, {
    message: "Either sender or transactionId must be provided",
  })
export type TransfersQuery = Partial<z.input<typeof TransfersQuerySchema>>

const TransferSchema = z.object({
  id: z.object({
    origin_chain: ChainSchema,
    origin_nonce: z.number().int().min(0),
  }),
  initialized: z.union([z.null(), TransactionSchema]),
  signed: z.union([z.null(), TransactionSchema]),
  finalised_on_near: z.union([z.null(), TransactionSchema]),
  finalised: z.union([z.null(), TransactionSchema]),
  claimed: z.union([z.null(), TransactionSchema]),
  transfer_message: TransferMessageSchema,
  updated_fee: z.array(TransactionSchema),
})

const ApiFeeResponseSchema = z.object({
  native_token_fee: safeBigInt(true),
  transferred_token_fee: safeBigInt(true).nullable(),
  usd_fee: z.number(),
})

const TransferStatusSchema = z.enum([
  "Initialized",
  "Signed",
  "FinalisedOnNear",
  "Finalised",
  "Claimed",
])

export type Transfer = z.infer<typeof TransferSchema>
export type ApiFeeResponse = z.infer<typeof ApiFeeResponseSchema>
export type TransferStatus = z.infer<typeof TransferStatusSchema>

interface ApiClientConfig {
  baseUrl?: string
}

class ApiError extends Error {
  constructor(
    message: string,
    public status?: number,
    public statusText?: string,
  ) {
    super(message)
    this.name = "ApiError"
  }
}

export class OmniBridgeAPI {
  private readonly baseUrl: string

  constructor(config: ApiClientConfig = {}) {
    this.baseUrl = config.baseUrl ?? this.getDefaultBaseUrl()
  }

  public getDefaultBaseUrl(): string {
    return getNetwork() === "testnet"
      ? "https://testnet.api.bridge.nearone.org"
      : "https://mainnet.api.bridge.nearone.org"
  }

  private async fetchWithValidation<T extends z.ZodType>(url: URL, schema: T): Promise<z.infer<T>> {
    const response = await fetch(url)

    if (response.status === 404) {
      throw new ApiError("Resource not found", response.status, response.statusText)
    }

    if (!response.ok) {
      throw new ApiError("API request failed", response.status, response.statusText)
    }

    const data = await response.json()
    return schema.parse(data)
  }

  private buildUrl(path: string, params: Record<string, string> = {}): URL {
    const url = new URL(`${this.baseUrl}${path}`)
    for (const [key, value] of Object.entries(params)) {
      url.searchParams.set(key, value)
    }
    return url
  }

  async getTransferStatus(originChain: Chain, originNonce: number): Promise<TransferStatus> {
    const url = this.buildUrl("/api/v1/transfers/transfer/status", {
      origin_chain: originChain,
      origin_nonce: originNonce.toString(),
    })
    return this.fetchWithValidation(url, TransferStatusSchema)
  }

  async getFee(
    sender: OmniAddress,
    recipient: OmniAddress,
    tokenAddress: OmniAddress,
  ): Promise<ApiFeeResponse> {
    const url = this.buildUrl("/api/v1/transfer-fee", {
      sender,
      recipient,
      token: tokenAddress,
    })
    return this.fetchWithValidation(url, ApiFeeResponseSchema)
  }

  async getTransfer(originChain: Chain, originNonce: number): Promise<Transfer> {
    const url = this.buildUrl("/api/v1/transfers/transfer", {
      // Removed trailing slash
      origin_chain: originChain,
      origin_nonce: originNonce.toString(),
    })
    return this.fetchWithValidation(url, TransferSchema)
  }

  async findOmniTransfers(query: TransfersQuery): Promise<Transfer[]> {
    const params = TransfersQuerySchema.parse(query)

    const urlParams: Record<string, string> = {
      offset: params.offset.toString(),
      limit: params.limit.toString(),
    }

    if (params.sender) urlParams.sender = params.sender
    if (params.transaction_id) urlParams.transaction_id = params.transaction_id

    const url = this.buildUrl("/api/v1/transfers", urlParams)
    return this.fetchWithValidation(url, z.array(TransferSchema))
  }
}
