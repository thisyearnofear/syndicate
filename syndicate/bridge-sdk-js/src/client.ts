import type { Provider as SolWallet } from "@coral-xyz/anchor"
import type { WalletSelector } from "@near-wallet-selector/core"
import { Wallet as EthWallet } from "ethers"
import { Account as NearAccount } from "near-api-js"
import { EvmBridgeClient } from "./clients/evm"
import { NearBridgeClient } from "./clients/near"
import { NearWalletSelectorBridgeClient } from "./clients/near-wallet-selector"
import { SolanaBridgeClient } from "./clients/solana"
import { addresses } from "./config"
import { ChainKind, type InitTransferEvent, type OmniTransferMessage } from "./types"
import { getBridgedToken, getChain } from "./utils"
import {
  getMinimumTransferableAmount,
  getTokenDecimals,
  verifyTransferAmount,
} from "./utils/decimals"

type Client =
  | EvmBridgeClient
  | NearBridgeClient
  | NearWalletSelectorBridgeClient
  | SolanaBridgeClient

// Type guards
export function isSolWallet(wallet: SolWallet | WalletSelector): wallet is SolWallet {
  return (
    wallet && typeof wallet === "object" && "publicKey" in wallet && "sendTransaction" in wallet
  )
}

export function isWalletSelector(wallet: SolWallet | WalletSelector): wallet is WalletSelector {
  return (
    wallet &&
    typeof wallet === "object" &&
    "wallet" in wallet &&
    typeof wallet.wallet === "function"
  )
}

/**
 * Validates and executes a cross-chain token transfer
 * @param wallet The wallet to use for the transfer
 * @param transfer The transfer details
 * @returns Promise resolving to transaction hash or InitTransferEvent
 * @throws If the transfer amount would be invalid after decimal normalization
 */
export async function omniTransfer(
  wallet: EthWallet | NearAccount | WalletSelector | SolWallet,
  transfer: OmniTransferMessage,
): Promise<string | InitTransferEvent> {
  // Get chain information
  const destChain = getChain(transfer.recipient)

  const sourceTokenAddress = transfer.tokenAddress
  const destTokenAddress = await getBridgedToken(transfer.tokenAddress, destChain)

  // Check if destination token address is null
  if (!destTokenAddress) {
    throw new Error(
      `Token ${transfer.tokenAddress} is not registered on the destination chain. Please deploy the token first.`,
    )
  }

  // Get token decimals
  const contractId = addresses.near // Use NEAR contract for decimal verification
  const sourceDecimals = await getTokenDecimals(contractId, sourceTokenAddress)
  const destinationDecimals = await getTokenDecimals(contractId, destTokenAddress)

  // Verify transfer amount will be valid after normalization
  const isValid = verifyTransferAmount(
    transfer.amount,
    transfer.fee,
    sourceDecimals.decimals,
    destinationDecimals.decimals,
  )

  if (!isValid) {
    // Get minimum amount
    const minAmount = getMinimumTransferableAmount(
      sourceDecimals.decimals,
      destinationDecimals.decimals,
    )
    throw new Error(
      `Transfer amount too small - would result in 0 after decimal normalization. Minimum transferable amount is ${minAmount}`,
    )
  }

  // Initialize appropriate client
  let client: Client | null = null

  if (wallet instanceof EthWallet) {
    client = new EvmBridgeClient(wallet, ChainKind.Eth)
  } else if (wallet instanceof NearAccount) {
    client = new NearBridgeClient(wallet)
  } else if (isSolWallet(wallet)) {
    client = new SolanaBridgeClient(wallet)
  } else if (isWalletSelector(wallet)) {
    client = new NearWalletSelectorBridgeClient(wallet)
  }

  if (!client) {
    throw new Error("Unsupported wallet type")
  }

  return await client.initTransfer(transfer)
}
