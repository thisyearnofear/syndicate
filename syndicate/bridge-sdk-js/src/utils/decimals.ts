import { getProviderByNetwork, view } from "@near-js/client"

import { addresses } from "../config"
import type { OmniAddress } from "../types"

export interface TokenDecimals {
  decimals: number
  origin_decimals: number
}

/**
 * Normalizes an amount from one decimal precision to another using BigInt math
 * @param amount The amount to normalize as a bigint
 * @param fromDecimals The source decimal precision
 * @param toDecimals The target decimal precision
 * @returns The normalized amount as a bigint
 */
export function normalizeAmount(amount: bigint, fromDecimals: number, toDecimals: number): bigint {
  if (fromDecimals === toDecimals) return amount

  if (fromDecimals > toDecimals) {
    // Scale down: Divide by power of 10
    const scale = 10n ** BigInt(fromDecimals - toDecimals)
    return amount / scale
    // biome-ignore lint/style/noUselessElse: Adds clarity
  } else {
    // Scale up: Multiply by power of 10
    const scale = 10n ** BigInt(toDecimals - fromDecimals)
    return amount * scale
  }
}

/**
 * Verifies if a transfer amount will be valid after normalization
 * @param amount The amount to transfer
 * @param fee The fee to be deducted
 * @param originDecimals The decimals of the token on the source chain
 * @param destinationDecimals The decimals of the token on the destination chain
 * @returns true if the normalized amount (minus fee) will be greater than 0
 */
export function verifyTransferAmount(
  amount: bigint,
  fee: bigint,
  originDecimals: number,
  destinationDecimals: number,
): boolean {
  try {
    // Use the minimum of origin and destination decimals for normalization
    const minDecimals = Math.min(originDecimals, destinationDecimals)

    // First normalize amount minus fee to the minimum decimals
    const normalizedAmount = normalizeAmount(amount - fee, originDecimals, minDecimals)

    // Check if amount minus fee is greater than 0
    return normalizedAmount > 0n
  } catch {
    // If we hit any math errors, the amount is effectively too small
    return false
  }
}

/**
 * Get the minimum number of decimals between origin and destination chains
 * @param contractId The NEAR contract to query
 * @param sourceToken The source token address
 * @param destinationToken The destination token address
 * @returns The minimum number of decimals
 */
export async function getMinimumDecimals(
  contractId: string,
  sourceToken: OmniAddress,
  destinationToken: OmniAddress,
): Promise<number> {
  const sourceDecimals = await getTokenDecimals(contractId, sourceToken)
  const destDecimals = await getTokenDecimals(contractId, destinationToken)

  return Math.min(sourceDecimals.decimals, destDecimals.decimals)
}

/**
 * Gets token decimals from the NEAR contract
 * @param contractId The NEAR contract ID to query
 * @param tokenAddress The Omni token address to check
 * @returns Promise resolving to the token's decimal information
 */
export async function getTokenDecimals(
  contractId: string,
  tokenAddress: OmniAddress,
): Promise<TokenDecimals> {
  const rpcProvider = getProviderByNetwork(addresses.network)
  return await view<TokenDecimals>({
    account: contractId,
    method: "get_token_decimals",
    args: {
      address: tokenAddress,
    },
    deps: { rpcProvider },
  })
}

/**
 * Gets the minimum transferable amount for a token pair accounting for decimal normalization
 * @param originDecimals The decimals of the source token
 * @param destinationDecimals The decimals of the destination token
 * @returns The minimum transferable amount as a bigint
 */
export function getMinimumTransferableAmount(
  originDecimals: number,
  destinationDecimals: number,
): bigint {
  // Start with 1 in destination decimal system
  let minAmount = 1n

  // If origin has more decimals, we need to scale up
  if (originDecimals > destinationDecimals) {
    minAmount = minAmount * 10n ** BigInt(originDecimals - destinationDecimals)
  }

  return minAmount
}
