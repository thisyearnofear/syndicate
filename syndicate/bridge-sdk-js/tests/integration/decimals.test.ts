import { afterAll, beforeAll, describe, it, vi } from "vitest"
import { setNetwork } from "../../src/config"
import { getTokenDecimals } from "../../src/utils/decimals"

describe.concurrent("getTokenDecimals integration", () => {
  setNetwork("testnet")

  beforeAll(() => {
    vi.spyOn(console, "error").mockImplementation(() => {})
    vi.spyOn(console, "debug").mockImplementation(() => {})
  })
  afterAll(() => {
    vi.restoreAllMocks()
  })

  it("fetches decimals for JLU token on Solana", async ({ expect }) => {
    const tokenAddress = "sol:rLFLkpdMZsVLWziDfz5WWqVgVnFbPdKicSNQcj9QVxL"
    const decimals = await getTokenDecimals("omni-locker.testnet", tokenAddress)

    // Verify response structure (these should not change)
    expect(decimals).toHaveProperty("decimals")
    expect(decimals).toHaveProperty("origin_decimals")
    expect(typeof decimals.decimals).toBe("number")
    expect(typeof decimals.origin_decimals).toBe("number")

    // Snapshot the actual values
    expect(decimals).toMatchSnapshot({
      decimals: 9,
      origin_decimals: 18,
    })
  }, 10000) // Increase timeout for RPC call

  it("handles invalid token addresses", async ({ expect }) => {
    const invalidAddress = "sol:invalid.testnet"
    await expect(getTokenDecimals("omni-locker.testnet", invalidAddress)).rejects.toMatchSnapshot()
  })
})
