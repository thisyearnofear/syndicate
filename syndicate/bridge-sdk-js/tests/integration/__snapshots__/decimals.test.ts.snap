// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTokenDecimals integration > fetches decimals for JLU token on Solana 1`] = `
{
  "decimals": 9,
  "origin_decimals": 18,
}
`;

exports[`getTokenDecimals integration > handles invalid token addresses 1`] = `[Error: Exceeded 2 providers to execute request]`;
